launcher/LaunchController.cpp

```diff
@@ -249,17 +257,30 @@ void LaunchController::login()
                 progDialog.execWithTask(task.get());
                 continue;
             }
-            // FIXME: this is missing - the meaning is that the account is queued for refresh and we should wait for that
-            /*
-            case AccountState::Queued: {
-                return;
-            }
-            */
             case AccountState::Expired: {
-                auto errorString = tr("The account has expired and needs to be logged into manually again.");
-                QMessageBox::warning(m_parentWidget, tr("Account refresh failed"), errorString, QMessageBox::StandardButton::Ok,
-                                     QMessageBox::StandardButton::Ok);
-                emitFailed(errorString);
+                auto errorString = tr("The account has expired and needs to be logged into manually. Press OK to log in again.");
+                auto button = QMessageBox::warning(m_parentWidget, tr("Account refresh failed"), errorString,
+                                                   QMessageBox::StandardButton::Ok | QMessageBox::StandardButton::Cancel,
+                                                   QMessageBox::StandardButton::Ok);
+                if (button == QMessageBox::StandardButton::Ok) {
+                    auto accounts = APPLICATION->accounts();
+                    bool isDefault = accounts->defaultAccount() == m_accountToUse;
+                    accounts->removeAccount(accounts->index(accounts->findAccountByProfileId(m_accountToUse->profileId())));
+                    if (m_accountToUse->isMSA()) {
+                        auto newAccount = MSALoginDialog::newAccount(
+                            m_parentWidget, tr("Please enter your Mojang account email and password to add your account."));
+                        accounts->addAccount(newAccount);
+                        if (isDefault) {
+                            accounts->setDefaultAccount(newAccount);
+                        }
+                        m_accountToUse = nullptr;
+                        decideAccount();
+                        continue;
+                    }
```

Comment:
I don't feel like this code belongs here. Ideally this should just call some function somewhere else that does all of this

Reply:
Not sure where else to put this code so for now I just extracted it as a separate function on the `LaunchController` I could not move it in AccountList as I would have polluted that with UI stuff. 
Do you have any suggestions about where you want it to be?